docs: Synthetics monitoring (#209)
elastic/observability-docs: docs/en/observability/synthetics-create-test.asciidoc

// REVIEWERS
// Should we add a note here about the `async` keyword / promises / link to mdn docs?
// https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous/Async_await[async/await].

[discrete]
[[synthetics-create-step]]
=== Create a `step`

Steps can be as simple or complex as you need them to be.
The {playwright-api-docs}[Playwright page API reference] will be your friend while writing tests.

A basic first step might simply load a page:

[source,js]
----
await page.goto('https://www.elastic.co'); <1>
----
<1> See the https://github.com/microsoft/playwright/blob/master/docs/api.md#pagegotourl-options[`page.goto` reference] for more information.

A more complex second step might wait for a page element to be selected,
and then ensure that it matches an expected value.
Consider the following HTML snippet:

[source,html]
----
<header class="header">
  <h1>todos</h1>
  <input class="new-todo"
    autofocus autocomplete="off"
    placeholder="What needs to be done?"
    v-model="newTodo"
    @keyup.enter="addTodo">
</header>
----

You can verify that `new-todo` class element contains the expected `placeholder` (input hint)
with the following test:

[source,js]
----

> const input = await page.$('input.new-todo'); <1>

Reviewer: should we have this inside step?

Author: After looking at this again, yes, I think you're right.

deepStrictEqual(await input.getAttribute('placeholder'), "What needs to be done?"); <2>
----
<1> Query the page for the `input.new-todo` selector.
See the https://github.com/microsoft/playwright/blob/master/docs/api.md#pageselector[`page.$` reference] for more information.
<2> Return the `placeholder` attribute for `new-todo` with https://developer.mozilla.org/en-US/docs/Web/API/Element/getAttribute[`getAttribute()`], and ensure it equals the provided text.
`deepStrictEqual` is provided by Node.js and tests for deep equality between the actual and
expected parameters.
See the https://nodejs.org/api/assert.html#assert_assert_deepstrictequal_actual_expected_message[Node.js API reference] for more information.

[discrete]
[[synthetics-sample-test]]
=== Sample Synthetic test

A complete example of a basic synthetic test looks like this:

[source,js]
----

> import { journey, step } from '@elastic/synthetics';

Reviewer: add `import { deepStrictEqual } from 'assert'; `

journey("Ensure placeholder is correct", async ({page}) => {
    step("Go to elastic.co", async () => {
      await page.goto('https://www.elastic.co');
    })
    step("Check placeholder text", async () => {
      const input = await page.$('input.new-todo');
      deepStrictEqual(await input.getAttribute('placeholder'), "What needs to be done?");
    })
});
----

[discrete]
[[synthetic-run-tests]]
== Running synthetic tests

// REVIEWERS:
// Should we doc required technologies?
// Node.js, npx, typescript, etc.?

As explained in the <<synthetics-quickstart,quickstart>>, there are two ways to run synthetic tests:

* Run as an inline journey -- copy/paste into `heartbeat.yml`
* Run a test suite -- import your tests into Heartbeat

Which option is right for you? That depends.

> If you only have a few tests, an inline journey might be the easiest workflow --

Reviewer: ```suggestion
If you want to invoke a single journey with few steps, an inline journey might be the easiest workflow --
```

you only need paste your tests into your `heartbeat.yml` file.
if you have a complex test suite, or your tests need to live with your application code,
you're likely better off setting up a synthetic test suite.

[discrete]
[[synthetics-inline-journey]]
=== Run an inline journey

The easiest way to run a synthetic test is by creating an inline journey.
The `journey` keyword isn't required, and access to variables like `page` and `params` is automatic.

Copy and paste your test steps into `heartbeat.yml`.

> Heartbeat will fork Node.js, schedule your tests, and start headless chrome.

Reviewer: ```suggestion
Heartbeat spawns a separate Node.js process, schedule your tests, and runs it on chromium browser.
```

You don't need to worry about anything else.

An example, `short.js`, is provided in the
https://github.com/elastic/synthetics/tree/master/examples/inline[elastic/synthetics] GitHub repository:

[source,js]
----
// test-homepage-hover.js
step("load homepage", async () => {
    await page.goto('https://www.elastic.co');
});
step("hover over products menu", async () => {
    await page.hover('css=[data-nav-item=products]');
});
----

To run this, or any other inline example locally, change into the directory of your test,

> and pipe the file contents to the `elastic-synthetics` command.

Reviewer: ```suggestion
and pipe the file contents to the `npx @elastic/synthetics` command.
```

For example:

[source,sh]
----

> cat examples/inline/short.js | npx synthetics --inline

Reviewer: ```suggestion
cat examples/inline/short.js | npx @elastic/synthetics --inline
```

----

If everything works as expected, you'll get the following response:

[source,sh]
----
Journey: inline
   ✓  Step: 'load homepage' succeeded (1831 ms)
   ✓  Step: 'hover over products menu' succeeded (97 ms)

 2 passed (2511 ms)
----

The script can then be copied into your in your `heartbeat.yml`:

[source,yml]
----
heartbeat.monitors:
- type: browser
  id: my-monitor
  name: My Monitor
  schedule: "@every 1m"
  script: |-
    step("load homepage", async () => {
        await page.goto('https://www.elastic.co');
    });
    step("hover over products menu", async () => {
        await page.hover('css=[data-nav-item=products]');
    });
----

That's it! You can either spin up Heartbeat yourself, or jump to <<synthetics-quickstart-step-three>>
of the Quickstart to use the provided Docker project template.

[discrete]
[[synthetics-test-suite]]
=== Run a test suite

If you have a suite of tests you'd like to implement, you can use Elastic synthetics as a library.
In this method, you use Docker to run both Heartbeat and `elastic-synthetics`.

// [discrete]
// [[synthetics-suite-install]]

> // ==== Step 1: Install `elastic-synthetics`

Reviewer: ```suggestion
// ==== Step 1: Install `@elastic/synthetics`
```

> If you haven't already, clone the https://github.com/elastic/synthetics[elastic/synthetics] repository

Reviewer: We published the package, so this shouldnt be necessary. Users can do `npm install -g @elastic/synthetics`

and install the project:

[source,sh]
----
git clone git@github.com:elastic/synthetics.git &&\
npm install
----

// [discrete]
// [[synthetics-suite-create]]
// ==== Step 2: Create your tests

Now it's time to write your tests:

. Create a new Node.js project.
. Create a `javascript` or `typescript` file that imports your tests.
All synthetic test files must use the `.journey.ts` or `.journey.js` file extension.

> . Call `run`, which is imported from `elastic-synthetics` and runs the CLI application.

Reviewer: Don't need this step one when invoked via CLI.

. Compile everything together.

At Elastic, we're fans of examples, so one is provided in `synthetics/examples/todo`.
Switch into the directory and install the application:

[source,sh]
----
cd examples/todos/ && npm install
----

From the root of the `synthetics` repo, you can now run the provided tests.
Any files matching the filename `*.journey.*` will be run.

[source,sh]
----

> npx elastic-synthetics examples/todos

Reviewer: ```suggestion
npx @elastic/synthetics examples/todos
```
